lib/libc/minimal/source/stdlib/qsort.c

```diff
@@ -80,7 +80,28 @@ static void heap_sort(void *base, int nmemb, size_t size, comp3_t comp, void *co
 	}
 }
 
-void qsort_r(void *base, size_t nmemb, size_t size, comp3_t comp, void *arg)
+static compar_fn_t qsort_compar_fn;
```

Comment:
nak. this is dangerous.

Reply:
How so, please explain.

Reply:
Saving the function pointer is not state that needs to be persisted. I suggest not saving a pointer to the comparison function

Reply:
In that case, I see no other way than to add an extra parameter to heap_sort, heapify and shift_down. The shift_down function can the check that parameter, and thereby determine which type of callback to call, with the correct number of parameters. How does that sound?

shift_down could then call, instead directly calling the compare function callback, a function similar to this:
> static int compare_cb(void *cmp_fn, bool cmp_fn_has_3, void *a, void *b, void *arg)
> {
>         int cmp;
>         if (cmp_fn_has_3) {
>                 cmp = ((compar_d_fn_t) cmp_fn)(a, b, arg);
>         } else {
>                 cmp = ((compar_fn_t) cmp_fn)(a, b);
>         }
>         return cmp;
> }
> 
Sorry for the poor formatting, I cannot get GitHub to format as multiline code, hope it is readable. 
Would that be a better option?